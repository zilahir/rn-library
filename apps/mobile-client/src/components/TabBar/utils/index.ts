export { default as icons } from "./icons";

export { onPress } from "./navigations";
