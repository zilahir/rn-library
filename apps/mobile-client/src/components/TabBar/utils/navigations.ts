import { BottomTabBarProps } from "@react-navigation/bottom-tabs";

interface INavigationProps {
  navigation: BottomTabBarProps["navigation"];
  targetKey: string;
}

/**
 *
 * @param root0
 * @param root0.navigation
 * @param root0.target
 * @param root0.isFocused
 * @param root0.targetKey
 * @param root0.targetName
 */
export function onPress({
  navigation,
  targetKey,
  isFocused,
  targetName,
}: INavigationProps & { isFocused: boolean; targetName: string }): void {
  const event = navigation.emit({
    type: "tabPress",
    target: targetKey,
    canPreventDefault: true,
  });

  if (!isFocused && !event.defaultPrevented) {
    navigation.navigate(targetName);
  }
}

/*
export function onLongPress({ navigation, targetKey }: INavigationProps): void {
  navigation.emit({
    type: "tabLongPress",
    target: targetKey,
  });
}
*/
