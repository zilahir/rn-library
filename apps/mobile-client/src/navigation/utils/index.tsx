import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";
import HomeScreen from "@app/screens/HomeScreen";
import ProfileScreen from "@app/screens/ProfileScreen";
import TrendingScreen from "@app/screens/TrendingScreen";
import LibraryScreen from "@app/screens/LibraryScreen";

type ScreenName = keyof BottomTabParamList;

interface BottomTabScreen {
  label?: ScreenName;
  name: ScreenName;
  tabComponent: ComponentProps<any>;
  hideLabel?: boolean;
  isFloating?: boolean;
}

interface IBottomTabScreens {
  screens: BottomTabScreen[];
  getBottomTabBarTabs: () => BottomTabScreen[];
  getNumberOfTabBarTabs: () => number;
}

export const bottomTabScreens: IBottomTabScreens = {
  screens: [
    {
      label: "Home",
      name: "Home",
      tabComponent: HomeScreen,
    },

    {
      label: "Profile",
      name: "Profile",
      tabComponent: ProfileScreen,
    },
    {
      label: "Trending",
      name: "Trending",
      tabComponent: TrendingScreen,
    },
    {
      label: "Library",
      name: "Library",
      tabComponent: LibraryScreen,
    },
  ],
  getBottomTabBarTabs: (): BottomTabScreen[] =>
    bottomTabScreens.screens.map((screen) => ({
      ...screen,
      hideLabel: !!screen.hideLabel,
      isFloating: !!screen.isFloating,
    })),

  getNumberOfTabBarTabs: (): number => bottomTabScreens.screens.length,
};
