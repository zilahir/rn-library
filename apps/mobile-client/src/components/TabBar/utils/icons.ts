import { Feather } from "@expo/vector-icons";
import { ComponentProps } from "react";
import { BottomTabParamList } from "@app/navigation/types";

const icons: Record<
  keyof BottomTabParamList,
  ComponentProps<typeof Feather>["name"]
> = {
  Home: "home",
  Profile: "user",
  Trending: "trending-up",
  Library: "book-open",
  Borrow: "book",
};

export default icons;
