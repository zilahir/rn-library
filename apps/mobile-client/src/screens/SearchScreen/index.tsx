import { ReactElement, useState } from "react";
import { View, StyleSheet, TouchableOpacity, ImageStyle } from "react-native";
import { Feather } from "@expo/vector-icons";
import { useTranslation } from "react-i18next";
import { useTheme } from "@shopify/restyle";
import { Book } from "pikkukirjasto-types/types/book";
import { FlatList } from "react-native-gesture-handler";
import Animated from "react-native-reanimated";
import withSafeAreaView from "@app/components/WithSafeAreaView";
import MainLayout from "@app/layouts/MainLayout";
import SectionTitle from "@app/components/common/SectionTitle";
import { Theme } from "@app/theme";
import { RootTabScreenProps } from "@app/navigation/types";
import Text from "@app/components/common/Text";
import SearchBar from "@app/components/SearchBar";
import useSearch from "@app/hooks/useSearch";

const styles = ({
  spacing,
  colors,
}: {
  spacing: Theme["spacing"];
  colors: Theme["colors"];
}): ReturnType<typeof StyleSheet.create> =>
  StyleSheet.create({
    searchResultsLoadingContainer: {
      justifyContent: "center",
      alignItems: "center",
      flex: 1,
    },
    searchResultRootContainer: {
      flex: 1,
    },
    placeHolder: {
      padding: spacing.s,
    },
    scrollViewContainer: {
      marginTop: spacing.l,
      flex: 1,
    },

    suggestionItem: {
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      marginBottom: spacing.s,
    },

    headerHelper: {
      marginTop: spacing.l * 2,
    },

    titleContainer: {
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
    },
    bookCoverImage: {
      width: 30,
      aspectRatio: 1 / 1.5,
    },
    searchResultItemSeparator: {
      height: 20,
    },
    searchIcon: {
      marginRight: spacing.m,
    },
    searchResultListContainer: {
      marginTop: spacing.m,
    },
    searchResultItemSeparatorBorder: {
      height: 3,
      backgroundColor: colors.orange100,
    },
  });

/**
 *
 * @param {object} root0 - props
 * @param {Book} root0.book - a single book object
 * @param {object} root0.navigation - navigation object passed from react-navigation
 * @returns {ReactElement} - react component
 * @description - renders a single suggestion item
 */
function SuggestionItem({
  book,
  navigation,
}: {
  book: Book;
  navigation: RootTabScreenProps<"Search">["navigation"];
}): ReactElement {
  const { spacing, colors } = useTheme<Theme>();

  const style = styles({ spacing, colors });

  return (
    <TouchableOpacity
      testID="suggestion-item"
      onPress={(): void =>
        navigation.navigate("Book", {
          book,
        })
      }
    >
      <Animated.View style={[style.suggestionItem]}>
        <Animated.Image
          source={{ uri: book.cover }}
          style={[style.bookCoverImage as ImageStyle]}
        />

        <View>
          <Text marginLeft="s" fontSize={14}>
            {book.title}
          </Text>
          <Text marginLeft="s" fontSize={12}>
            {book.author}
          </Text>
        </View>
      </Animated.View>
    </TouchableOpacity>
  );
}

/**
 *
 * @param {object} root0 - props
 * @param {object} root0.navigation - navigation object passed from react-navigation
 * @param {object} root0.route - route object passed from react-navigation
 * @returns {ReactElement} - react component
 * @description - renders a search screen
 */
function SearchScreen({
  navigation,
  route,
}: RootTabScreenProps<"Search">): ReactElement {
  const [searchResult, setSearchResult] = useState<Book[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>("");

  const { handleSorting } = useSearch();

  const { t } = useTranslation();

  /**
   *
   * @param {string} queryString - query string
   * @returns {Promise<Book[]>} - array of books
   * @description - handles search
   */
  // istanbul ignore next
  async function handleSearch(queryString: string): Promise<void> {
    setSearchTerm(queryString);
    const result = await handleSorting(queryString);
    setSearchResult(result);
  }

  const { spacing, colors } = useTheme<Theme>();
  const style = styles({ spacing, colors });

  return (
    <MainLayout>
      <View style={[style.headerHelper]}>
        <SearchBar
          searchTerm={searchTerm}
          setSearchTerm={(text): Promise<void> => handleSearch(text)}
          route={route}
          navigation={navigation}
        />
      </View>
      <View>
        <View style={[style.titleContainer]}>
          <Feather
            style={style.searchIcon}
            size={20}
            name="search"
            color={colors.gray}
          />
          <SectionTitle title={t("searchScreen.suggestions")} />
        </View>
        <FlatList
          data={searchResult}
          keyExtractor={(item): string => item._id}
          ItemSeparatorComponent={(): ReactElement => (
            <View style={[style.searchResultItemSeparator]}>
              <View style={[style.searchResultItemSeparatorBorder]} />
            </View>
          )}
          contentContainerStyle={[style.searchResultListContainer]}
          renderItem={({ item }): ReactElement => (
            <SuggestionItem
              navigation={navigation}
              book={item}
              key={item._id}
            />
          )}
        />
      </View>
    </MainLayout>
  );
}

export default withSafeAreaView(SearchScreen);
