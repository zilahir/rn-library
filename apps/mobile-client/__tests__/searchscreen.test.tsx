/* eslint-disable react/jsx-props-no-spreading */
import { fireEvent } from "@testing-library/react-native";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import React from "react";
import SearchScreen from "@app/screens/SearchScreen";

import { render } from "./utils/wrapper";

const demoBooks = [
  {
    isbn: "demo isbn",
    author: "Terrell Stroman",
    title: "dedico adhuc tripudio",
    cover: "https://picsum.photos/seed/9YjBfGro/640/480",
  },
  {
    isbn: "demo isbn",
    author: "Eleanor Walter DDS",
    title: "statim aiunt defaeco",
    cover: "https://loremflickr.com/640/480?lock=1138305395589120",
  },
  {
    isbn: "demo isbn",
    author: "Thelma Mills",
    title: "terga vitae",
    cover: "https://loremflickr.com/640/480?lock=2883154006245376",
  },
  {
    isbn: "demo isbn",
    author: "Alberta Klein",
    title: "vulnus ratione",
    cover: "https://picsum.photos/seed/D5B0dZ5/640/480",
  },
  {
    isbn: "demo isbn",
    author: "Milton Ledner",
    title: "carus explicabo ustilo",
    cover: "https://picsum.photos/seed/22MXxVPpUr/640/480",
  },
  {
    isbn: "demo isbn",
    author: "Jon Marks",
    title: "accedo corporis",
    cover: "https://loremflickr.com/640/480?lock=4790681306398720",
  },
  {
    isbn: "demo isbn",
    author: "Saul Cruickshank-Ziemann",
    title: "accusator debilito",
    cover: "https://loremflickr.com/640/480?lock=4005968851501056",
  },
  {
    isbn: "demo isbn",
    author: "Abraham Hermann",
    title: "voluptatum vergo est",
    cover: "https://picsum.photos/seed/ZOBIznvngn/640/480",
  },
  {
    isbn: "demo isbn",
    author: "Drew Larson",
    title: "vilitas",
    cover: "https://picsum.photos/seed/cRcFs59PH/640/480",
  },
  {
    isbn: "demo isbn",
    author: "Mr. Harvey VonRueden",
    title: "stips amitto thalassinus",
    cover: "https://loremflickr.com/640/480?lock=2325913328943104",
  },
];

const props = {
  navigation: {
    navigate: jest.fn(),
  },
  route: "",
  searchTerm: "",
  setSearchTerm: jest.fn(),
  book: {
    ...demoBooks[0],
  },
} as any;

describe("SearhScreen", () => {
  beforeAll(() => {
    jest.mock("@app/queries/book", () => ({
      useBooks: jest.fn(() => {
        return {
          data: [...demoBooks],
          isLoading: false,
        };
      }),
    }));
  });
  it("renders correctly", () => {
    const { toJSON } = render(
      <QueryClientProvider client={new QueryClient()}>
        <SearchScreen {...props} />
      </QueryClientProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("navigation is called", () => {
    jest.spyOn(React, "useState").mockReturnValueOnce([demoBooks, jest.fn()]);

    const { queryAllByTestId } = render(
      <QueryClientProvider client={new QueryClient()}>
        <SearchScreen {...props} />
      </QueryClientProvider>
    );
    fireEvent.press(queryAllByTestId("suggestion-item")[0]);
    expect(props.navigation.navigate).toHaveBeenCalled();
  });
});
