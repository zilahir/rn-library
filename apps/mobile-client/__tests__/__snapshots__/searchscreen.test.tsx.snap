// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearhScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "backgroundColor": "#fff",
        "borderBottomLeftRadius": 25,
        "borderBottomRightRadius": 25,
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
          "marginHorizontal": 24,
          "marginVertical": 16,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "marginTop": 48,
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginBottom": 16,
            "opacity": 1,
          }
        }
        testID="search-bar"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "rgba(247, 247, 247, 1)",
                "borderRadius": 10,
                "display": "flex",
                "flexDirection": "row",
                "height": 40,
                "paddingHorizontal": 8,
              },
            ]
          }
        >
          <Text />
          <TextInput
            editable={false}
            onChangeText={[Function]}
            placeholder="Search"
            placeholderTextColor="rgba(247, 247, 247, 1)"
            style={
              [
                {
                  "color": "#828282",
                  "flex": 1,
                  "fontSize": 14,
                  "paddingLeft": 10,
                },
              ]
            }
            testID="text-input"
            value=""
          />
        </View>
      </View>
    </View>
    <View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text />
        <View>
          <Text
            style={
              [
                {
                  "color": "#333333",
                  "fontSize": 22,
                  "fontWeight": 500,
                },
                [
                  undefined,
                ],
              ]
            }
          >
            searchScreen.suggestions
          </Text>
        </View>
      </View>
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        collapsable={false}
        contentContainerStyle={
          [
            {
              "marginTop": 16,
            },
          ]
        }
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        handlerTag={1}
        handlerType="NativeViewGestureHandler"
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        renderScrollComponent={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
        waitFor={
          [
            {
              "current": null,
            },
            {
              "current": null,
            },
          ]
        }
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</RCTSafeAreaView>
`;
